singularity-opac/neutrinos/mean_opacity_neutrinos.hpp

```diff
@@ -0,0 +1,212 @@
+              kappaRosselandDenom +=
+                  opac.DThermalDistributionOfTNuDT(T, type, nu) * nu * dlnu;
+            }
+
+            // Trapezoidal rule
+            const Real nu0 = fromLog_(lnuMin);
+            const Real nu1 = fromLog_(lnuMax);
+            kappaPlanckNum -=
+                0.5 * 1. / rho *
+                (opac.AbsorptionCoefficient(rho, T, Ye, type, nu0, lambda) *
+                     opac.ThermalDistributionOfTNu(T, type, nu0) * nu0 +
+                 opac.AbsorptionCoefficient(rho, T, Ye, type, nu1, lambda) *
+                     opac.ThermalDistributionOfTNu(T, type, nu1) * nu1) *
+                dlnu;
+            kappaPlanckDenom -=
+                0.5 *
+                (opac.ThermalDistributionOfTNu(T, type, nu0) * nu0 +
+                 opac.ThermalDistributionOfTNu(T, type, nu1) * nu1) *
+                dlnu;
+            kappaRosselandNum -=
+                0.5 * rho *
+                (1. /
+                     opac.AbsorptionCoefficient(rho, T, Ye, type, nu0, lambda) *
+                     opac.DThermalDistributionOfTNuDT(T, type, nu0) * nu0 +
+                 1. /
+                     opac.AbsorptionCoefficient(rho, T, Ye, type, nu1, lambda) *
+                     opac.DThermalDistributionOfTNuDT(T, type, nu1) * nu1) *
+                dlnu;
+            kappaRosselandDenom -=
+                0.5 *
+                (opac.DThermalDistributionOfTNuDT(T, type, nu0) * nu0 +
+                 opac.DThermalDistributionOfTNuDT(T, type, nu1) * nu1) *
+                dlnu;
+
+            Real lkappaPlanck = toLog_(kappaPlanckNum / kappaPlanckDenom);
+            Real lkappaRosseland =
+                toLog_(1. / (kappaRosselandNum / kappaRosselandDenom));
+            lkappaPlanck_(iRho, iT, iYe, idx) = lkappaPlanck;
+            lkappaRosseland_(iRho, iT, iYe, idx) = lkappaRosseland;
+          }
+        }
+      }
+    }
+  }
+
+#ifdef SPINER_USE_HDF
+  SpinerOpacity(const std::string &filename)
+      : filename_(filename.c_str()), memoryStatus_(impl::DataStatus::OnHost) {
+    herr_t status = H5_SUCCESS;
+    hid_t file = H5Fopen(filename.c_str(), H5F_ACC_RDONLY, H5P_DEFAULT);
+    status += lkappaPlanck_.loadHDF(file, SP5::Opac::PlanckMeanOpacity);
+    status += lkappaRosseland_.loadHDF(file, SP5::Opac::RosselandMeanOpacity);
+    status += H5Fclose(file);
+
+    if (status != H5_SUCCESS) {
+      OPAC_ERROR("neutrinos::MeanOpacity: HDF5 error\n");
+    }
+  }
+
+  void Save(const std::string &filename) const {
+    herr_t status = H5_SUCCESS;
+    hid_t file =
+        H5Fcreate(filename.c_str(), H5F_ACC_TRUNC, H5P_DEFAULT, H5P_DEFAULT);
+    status += lkappaPlanck_.saveHDF(file, SP5::Opac::PlanckMeanOpacity);
+    status += lkappaRosseland_.saveHDF(file, SP5::Opac::RosselandMeanOpacity);
+    status += H5Fclose(file);
+
+    if (status != H5_SUCCESS) {
+      OPAC_ERROR("neutrinos::MeanOpacity: HDF5 error\n");
+    }
+  }
```

Comment:
Nice. Caching mean opacities in file seems useful.

Reply:
:+1: